proposals/solid-notifications.md

```diff
@@ -0,0 +1,270 @@
+    "type": ["WebSocketSubscription2021", "WebHookSubscription2021"],
+    "features": ["state", "rate"]
+}
+```
+
+#### Response
+
+```http
+HTTP/2
+Content-Type: application/ld+json
+
+{
+    "@context": ["https://www.w3.org/ns/solid/notification/v1"],
+    "type": "WebSocketSubscription2021",
+    "endpoint": "https://websocket.example/subscription",
+    "features": ["state", "rate", "expiration", "feature1", "feature2"]
+}
+```
+
+**Note**: the features array may not be needed as part of this negotiation.
+
+### Notification Subscription
+
+The details of subscribing to a particular endpoint depend on the technology used, but all will follow a similar pattern. For WebSockets, this involves sending an authenticated subscription request to the endpoint retrieved from the gateway service.
+
+The client sends a JSON-LD payload to this endpoint via POST. The only required fields in this interaction are type and topic. The type field MUST contain the type of subscription being requested. The topic field MUST contain the URL of the resource a client wishes to subscribe to changes.
+
+All other fields are defined by the particular subscription type. Some common features are described in the Features section of this document.
+
+An example POST request using a DPoP bound access token is below:
+
+```http
+POST /subscription
+Authorization: DPoP <token>
+DPoP: <proof>
+Content-Type: application/ld+json
+
+{
+    "@context": ["https://www.w3.org/ns/solid/notification/v1"],
+    "type": "WebSocketSubscription2021",
+    "topic": "https://pod.example/resource",
+    "state": "opaque-state",
+    "expiration": "2021-09-21T12:37:15Z",
+    "rate": "PT10s"
+}
+```
+
+A successful response will contain a URL that can be used directly with a JavaScript client
+
+```http
+HTTP/2
+Content-Type: application/ld+json
+
+{
+    "@context": "https://www.w3.org/ns/solid/notification/v1",
+    "type": "WebSocketSubscription2021",
+    "endpoint": "wss://websocket.example/?auth=Ys3KiUq"
+}
+```
+
+In JavaScript, a client can use the data in the response to establish a connection to the WebSocket endpoint.
+
+```javascript
+const ws = new WebSocket(endpoint, type)
+```
+
+A client will define how to handle events, especially the onclose and onmessage events.
+
+```javascript
+ws.onclose(evt => reconnect(evt));
+ws.onmessage(evt => console.log("Message received: ", evt))
+```
+
+## Authentication and Authorization
+
+The Subscription API requires authorization. This document does not define the specific technology used to authorize requests; rather, it defers to the Solid Protocol sections on Authentication and Authorization. It is out of scope for this document to describe how a client fetches an authorization token. Solid-OIDC is one example of an authentication mechanism that could be used with Solid Notifications.
+
+Resource-based Access controls MUST be enforced for every subscription. A client MUST be permitted to read the resource to which it subscribes.
```

Comment:
```suggestion
Resource-based Access controls MUST be enforced for every subscription. A client MUST be permitted to read the resources to which it subscribes.
```

Suggest pluralizing, since we should be able to subscribe eventually to topics that apply to multiple resources.

Reply:
The challenge of subscribing to multiple resources on a single channel is that other features of this interaction fall apart.

For instance, if `state` refers to the last state of a resource, what does it mean for multiple resources?
If a client accepts updates at only a certain rate, what happens if two messages appear from different resources? Is one dropped?

At the underlying TCP layer (when HTTP/2 is used), all of the interactions occur over a single TCP socket, so there isn't additional network overhead when opening multiple such connections.